<h1><center>Laboratorio 6: Optimización de modelos 🧪</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos - Primavera 2025</strong></center>

### Cuerpo Docente:

- Profesores: Diego Cortez, Gabriel Iturra
- Auxiliares: Melanie Peña, Valentina Rojas
- Ayudantes: Nicolás Cabello, Cristopher Urbina

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1: María Jesús Valenzuela
- Nombre de alumno 2: Matías Zúñiga

Reglas:

- **Grupos de 2 personas**
- Fecha de entrega: Entregas Martes a las 23:59.
- Instrucciones del lab el viernes a las 16:15 en formato online. Asistencia no es obligatoria, pero se recomienda **fuertemente** asistir.
- <u>Prohibidas las copias</u>. Cualquier intento de copia será debidamente penalizado con el reglamento de la escuela.
- Tienen que subir el laboratorio a u-cursos y a su repositorio de github. Labs que no estén en u-cursos no serán revisados. Recuerden que el repositorio también tiene nota.
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Pueden usar cualquier material del curso que estimen conveniente.

### **Link de repositorio de GitHub:** [Insertar Repositorio](https://github.com/jesuvalenzuela/NanSquad)

### Temas a tratar

- Predicción de demanda usando `xgboost`
- Búsqueda del modelo óptimo de clasificación usando `optuna`
- Uso de pipelines.


### Reglas:

- **Grupos de 2 personas**
- Fecha de entrega: 6 días de plazo con descuento de 1 punto por día. Entregas Martes a las 23:59.
- Instrucciones del lab el viernes a las 16:15 en formato online. Asistencia no es obligatoria, pero se recomienda fuertemente asistir.
- <u>Prohibidas las copias</u>. Cualquier intento de copia será debidamente penalizado con el reglamento de la escuela.
- Tienen que subir el laboratorio a u-cursos y a su repositorio de github. Labs que no estén en u-cursos no serán revisados. Recuerden que el repositorio también tiene nota.
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Pueden usar cualquier material del curso que estimen conveniente.

El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

# Importamos librerias útiles

In [2]:
# !pip install -qq xgboost optuna
# !pip install optuna-integration[xgboost]

import pandas as pd
import numpy as np
from datetime import datetime

from sklearn import set_config
set_config(transform_output="pandas")

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import FunctionTransformer, OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.pipeline import Pipeline
from sklearn.dummy import DummyRegressor
from sklearn.metrics import mean_absolute_error

from xgboost import XGBRegressor

import joblib

import optuna
from optuna.samplers import TPESampler
optuna.logging.set_verbosity(optuna.logging.WARNING)
import time

# El emprendimiento de Fiu

Tras liderar de manera exitosa la implementación de un proyecto de ciencia de datos para caracterizar los datos generados en Santiago 2023, el misterioso corpóreo **Fiu** se anima y decide levantar su propio negocio de consultoría en machine learning. Tras varias e intensas negociaciones, Fiu logra encontrar su *primera chamba*: predecir la demanda (cantidad de venta) de una famosa productora de bebidas de calibre mundial. Al ver el gran potencial y talento que usted ha demostrado en el campo de la ciencia de datos, Fiu lo contrata como data scientist para que forme parte de su nuevo emprendimiento.

Para este laboratorio deben trabajar con los datos `sales.csv` subidos a u-cursos, el cual contiene una muestra de ventas de la empresa para diferentes productos en un determinado tiempo.

Para comenzar, cargue el dataset señalado y visualice a través de un `.head` los atributos que posee el dataset.

<i><p align="center">Fiu siendo felicitado por su excelente desempeño en el proyecto de caracterización de datos</p></i>
<p align="center">
  <img src="https://media-front.elmostrador.cl/2023/09/A_UNO_1506411_2440e.jpg">
</p>

In [3]:
# Importar datos
df = pd.read_csv('sales.csv', 
                 parse_dates = ['date'],
                 date_format="%d/%m/%y", 
                 index_col='id')

df.head()

date    city       lat      long     pop    shop        brand  \
id                                                                       
0  2012-01-31  Athens  37.97945  23.71622  672130  shop_1  kinder-cola   
1  2012-01-31  Athens  37.97945  23.71622  672130  shop_1  kinder-cola   
2  2012-01-31  Athens  37.97945  23.71622  672130  shop_1  kinder-cola   
3  2012-01-31  Athens  37.97945  23.71622  672130  shop_1   adult-cola   
4  2012-01-31  Athens  37.97945  23.71622  672130  shop_1   adult-cola   

   container capacity  price  quantity  
id                                      
0      glass    500ml   0.96     13280  
1    plastic    1.5lt   2.86      6727  
2        can    330ml   0.87      9848  
3      glass    500ml   1.00     20050  
4        can    330ml   0.39     25696

## 1 Generando un Baseline (5 puntos)

<p align="center">
  <img src="https://media.tenor.com/O-lan6TkadUAAAAC/what-i-wnna-do-after-a-baseline.gif">
</p>

Antes de entrenar un algoritmo, usted recuerda los apuntes de su magíster en ciencia de datos y recuerda que debe seguir una serie de *buenas prácticas* para entrenar correcta y debidamente su modelo. Después de un par de vueltas, llega a las siguientes tareas:

1. Separe los datos en conjuntos de train (70%), validation (20%) y test (10%). Fije una semilla para controlar la aleatoriedad. [0.5 puntos]
2. Implemente un `FunctionTransformer` para extraer el día, mes y año de la variable `date`. Guarde estas variables en el formato categorical de pandas. [1 punto]
3. Implemente un `ColumnTransformer` para procesar de manera adecuada los datos numéricos y categóricos. Use `OneHotEncoder` para las variables categóricas. `Nota:` Utilice el método `.set_output(transform='pandas')` para obtener un DataFrame como salida del `ColumnTransformer` [1 punto]
4. Guarde los pasos anteriores en un `Pipeline`, dejando como último paso el regresor `DummyRegressor` para generar predicciones en base a promedios. [0.5 punto]
5. Entrene el pipeline anterior y reporte la métrica `mean_absolute_error` sobre los datos de validación. ¿Cómo se interpreta esta métrica para el contexto del negocio? [0.5 puntos]
6. Finalmente, vuelva a entrenar el `Pipeline` pero esta vez usando `XGBRegressor` como modelo **utilizando los parámetros por default**. ¿Cómo cambia el MAE al implementar este algoritmo? ¿Es mejor o peor que el `DummyRegressor`? [1 punto]
7. Guarde ambos modelos en un archivo .pkl (uno cada uno) [0.5 puntos]

In [4]:
# Inserte su código acá

# === 1. Separar los datos en conjuntos de train (70%), validation (20%) y test (10%).
X = df.drop(columns=['quantity'])
y = df['quantity']

# Primero separamos conjunto de test
X_train_val, X_test, y_train_val, y_test = train_test_split(
    X, y, test_size=0.1, random_state=42)

# Luego separamos los datos restantes en entrenamiento y validación
val_size = 0.2 / 0.9        # proporción de datos restantes que representan el 20% del total
X_train, X_val, y_train, y_val = train_test_split(
    X_train_val, y_train_val, test_size=val_size, random_state=42)


# === 2. Implementar un `FunctionTransformer` para extraer el día, mes y año de la variable `date`.

def extract_date_features(X):
    """
    Extrae día, mes y año de la columna 'date' y los convierte a categorical.
    """
    X_copy = X.copy()
    
    # Asegurar formato datetime
    if not pd.api.types.is_datetime64_any_dtype(X_copy['date']):
        X_copy['date'] = pd.to_datetime(X_copy['date'])
    
    # Extraer características temporales
    X_copy['day'] = X_copy['date'].dt.day.astype('category')
    X_copy['month'] = X_copy['date'].dt.month.astype('category')
    X_copy['year'] = X_copy['date'].dt.year.astype('category')
    
    # Eliminar columna original
    X_copy = X_copy.drop(columns=['date'])
    
    return X_copy


# Crear FunctionTransformer
date_transformer = FunctionTransformer(extract_date_features)


# === 3. Implementar un `ColumnTransformer` para procesar de manera adecuada los datos numéricos y categóricos.

class AutoColumnTransformer(BaseEstimator, TransformerMixin):
    """
    ColumnTransformer que identifica automáticamente variables numéricas y categóricas.
    
    Durante fit(), identifica los tipos de columnas y crea internamente un 
    ColumnTransformer apropiado. Esto evita tener que especificar columnas manualmente.
    """
    
    def __init__(self, min_frequency=None):
        """
        Inicializa el transformador.
        """
        self.min_frequency = min_frequency
        self.column_transformer_ = None
    
    def fit(self, X, y=None):
        """
        Identifica columnas numéricas y categóricas, y crea el ColumnTransformer.
        """
        # Identificar tipos de columnas
        numeric_features = X.select_dtypes(include='number').columns.tolist()
        categorical_features = X.select_dtypes(include=['object', 'category']).columns.tolist()
        
        # Crear ColumnTransformer
        self.column_transformer_ = ColumnTransformer(
            transformers=[
                ('num', StandardScaler(), numeric_features),
                ('cat', OneHotEncoder(
                    drop='first',
                    sparse_output=False,
                    min_frequency=self.min_frequency        # Para la sección de optimización con optuna
                    ), categorical_features)
                ],
            verbose_feature_names_out=False
            )
        
        # Configurar salida como DataFrame
        self.column_transformer_.set_output(transform='pandas')
        
        # Ajustar el ColumnTransformer
        self.column_transformer_.fit(X, y)
        
        return self
    
    def transform(self, X):
        """
        Aplica la transformación.
        """
        return self.column_transformer_.transform(X)

# Instanciar el transformador    
col_transformer = AutoColumnTransformer()

# === 4. Guardar los pasos anteriores en un `Pipeline`
pipeline = Pipeline(steps=[
    ('date_features', date_transformer),
    ('preprocessor', col_transformer),
    ('regressor', DummyRegressor(strategy='mean'))
])

# === 5. Entrenar el pipeline y reportar la métrica `mean_absolute_error` sobre los datos de validación.

pipeline.fit(X_train, y_train)
y_val_pred = pipeline.predict(X_val)

print(f"MAE validación Dummy = {mean_absolute_error(y_val, y_val_pred):.4f}")

# === 6. Finalmente, vuelva a entrenar el `Pipeline` pero esta vez usando `XGBRegressor` como modelo **utilizando los parámetros por default**. 

pipeline_xgb = Pipeline(steps=[
    ('date_features', date_transformer),
    ('preprocessor', col_transformer),
    ('regressor', XGBRegressor(random_state=42))
])

pipeline_xgb.fit(X_train, y_train)
y_val_pred_xgb = pipeline_xgb.predict(X_val)

print(f"MAE validación XGB = {mean_absolute_error(y_val, y_val_pred_xgb):.4f}")

# === 7. Guardar ambos modelos en un archivo .pkl

# Guardar el pipeline con DummyRegressor
joblib.dump(pipeline, 'pipeline_dummy.pkl')

# Guardar el pipeline con XGBRegressor
joblib.dump(pipeline_xgb, 'pipeline_xgb.pkl')

print("Modelos guardados")

MAE validación Dummy = 13543.9614
MAE validación XGB = 2412.0123
Modelos guardados


5. **¿Cómo se interpreta esta métrica para el contexto del negocio?**

Para el contexto del negocio, el mean absolute error (MAE), indica, la diferencia promedio entre las ventas estimadas y las reales. De manera más literal, el MAE indica la media de los errores absolutos de la estimación. 

6. **¿Cómo cambia el MAE al implementar este algoritmo? ¿Es mejor o peor que el `DummyRegressor`?**

El MAE disminuye considerablemente con XGBRegressor, en comparación al modelo Dummy. Esto significa una mejora significativa en el desempeño predictivo.


## 2. Forzando relaciones entre parámetros con XGBoost (10 puntos)

<p align="center">
  <img src="https://64.media.tumblr.com/14cc45f9610a6ee341a45fd0d68f4dde/20d11b36022bca7b-bf/s640x960/67ab1db12ff73a530f649ac455c000945d99c0d6.gif">
</p>

Un colega aficionado a la economía le *sopla* que la demanda guarda una relación inversa con el precio del producto. Motivado para impresionar al querido corpóreo, se propone hacer uso de esta información para mejorar su modelo realizando las siguientes tareas:

1. Vuelva a entrenar el `Pipeline` con `XGBRegressor`, pero esta vez forzando una relación monótona negativa entre el precio y la cantidad. Para aplicar esta restricción apóyese en la siguiente <a href = https://xgboost.readthedocs.io/en/stable/tutorials/monotonic.html>documentación</a>. [6 puntos]

>Hint 1: Para implementar el constraint se le sugiere hacerlo especificando el nombre de la variable. De ser así, probablemente le sea útil **mantener el formato de pandas** antes del step de entrenamiento.

>Hint 2: Puede obtener el nombre de las columnas en el paso anterior al modelo regresor mediante el método `.get_feature_names_out()`

2. Luego, vuelva a reportar el `MAE` sobre el conjunto de validación. [1 puntos]

3. ¿Cómo cambia el error al incluir esta relación? ¿Tenía razón su amigo? [2 puntos]

4. Guarde su modelo en un archivo .pkl [1 punto]

In [5]:
# Inserte su código acá

# === 1. Pipeline con restricciones monotónicas
pipeline_xgb2 = Pipeline(steps=[
    ('date_features', date_transformer),
    ('preprocessor', col_transformer),
    ('regressor', XGBRegressor(
        monotone_constraints={'price': -1},     # relación monótona negativa con precio
        random_state=42
    ))
])

# Entrenar y predecir
pipeline_xgb2.fit(X_train, y_train)
y_val_pred_xgb2 = pipeline_xgb2.predict(X_val)

# === 2. Reportar MAE
print(f"MAE validación XGB monótono = {mean_absolute_error(y_val, y_val_pred_xgb2):.4f}")

# === 4. Guardar modelo
joblib.dump(pipeline_xgb2, 'pipeline_xgb_monotono.pkl')

print("Modelo guardado")

MAE validación XGB monótono = 2555.8156
Modelo guardado


3. **¿Cómo cambia el error al incluir esta relación? ¿Tenía razón su amigo?**

El error aumenta al incluir esta relación, por lo que el amigo no tenía razón. En general, este tipo de restricción introduce sesgos en el entrenamiento, que pueden perjudicar el desempeño del modelo.

## 3. Optimización de Hiperparámetros con Optuna (20 puntos)

<p align="center">
  <img src="https://media.tenor.com/fmNdyGN4z5kAAAAi/hacking-lucy.gif">
</p>

Luego de presentarle sus resultados, Fiu le pregunta si es posible mejorar *aun más* su modelo. En particular, le comenta de la optimización de hiperparámetros con metodologías bayesianas a través del paquete `optuna`. Como usted es un aficionado al entrenamiento de modelos de ML, se propone implementar la descabellada idea de su jefe.

A partir de la mejor configuración obtenida en la sección anterior, utilice `optuna` para optimizar sus hiperparámetros. En particular, se pide que su optimización considere lo siguiente:

- Fijar una semilla en las instancias necesarias para garantizar la reproducibilidad de resultados
- Utilice `TPESampler` como método de muestreo
- De `XGBRegressor`, optimice los siguientes hiperparámetros:
    - `learning_rate` buscando valores flotantes en el rango (0.001, 0.1)
    - `n_estimators` buscando valores enteros en el rango (50, 1000)
    - `max_depth` buscando valores enteros en el rango (3, 10)
    - `max_leaves` buscando valores enteros en el rango (0, 100)
    - `min_child_weight` buscando valores enteros en el rango (1, 5)
    - `reg_alpha` buscando valores flotantes en el rango (0, 1)
    - `reg_lambda` buscando valores flotantes en el rango (0, 1)
- De `OneHotEncoder`, optimice el hiperparámetro `min_frequency` buscando el mejor valor flotante en el rango (0.0, 1.0)

Para ello se pide los siguientes pasos:
1. Implemente una función `objective()` que permita minimizar el `MAE` en el conjunto de validación. Use el método `.set_user_attr()` para almacenar el mejor pipeline entrenado. [10 puntos]
2. Fije el tiempo de entrenamiento a 5 minutos. [1 punto]
3. Optimizar el modelo y reportar el número de *trials*, el `MAE` y los mejores hiperparámetros encontrados. ¿Cómo cambian sus resultados con respecto a la sección anterior? ¿A qué se puede deber esto? [3 puntos]
4. Explique cada hiperparámetro y su rol en el modelo. ¿Hacen sentido los rangos de optimización indicados? [5 puntos]
5. Guardar su modelo en un archivo .pkl [1 punto]

In [6]:
# Inserte su código acá

# === 1. Implementar una función `objective()` que permita minimizar el `MAE` en el conjunto de validación.

# No consideramos 'monotone_constraints', ya que NO es "la mejor configuración obtenida en la sección anterior"

def objective(trial):
    """
    Función objetivo para optimización con Optuna.
    Minimiza el MAE en el conjunto de validación.
    """

    # Parámetros XBGRegressor
    params_xgb = {
            'learning_rate': trial.suggest_float('learning_rate', 0.001, 0.1),
            'n_estimators': trial.suggest_int('n_estimators', 50, 1000),
            'max_depth': trial.suggest_int('max_depth', 3, 10),
            'max_leaves': trial.suggest_int('max_leaves', 0, 100),
            'min_child_weight': trial.suggest_int('min_child_weight', 1, 5),
            'reg_alpha': trial.suggest_float('reg_alpha', 0, 1),
            'reg_lambda': trial.suggest_float('reg_lambda', 0, 1),
            'random_state': 42
        }

    # Parámetros OneHotEncoder
    min_frequency = trial.suggest_float('min_frequency', 0.0, 1.0)

    # Pipeline
    pipeline = Pipeline(steps=[
        ('date_features', date_transformer),
        ('preprocessor', AutoColumnTransformer(min_frequency=min_frequency)),
        ('regressor', XGBRegressor(**params_xgb))
    ])

    # Entrenar el pipeline
    pipeline.fit(X_train, y_train)
    
    # Predecir en validación
    y_pred_val = pipeline.predict(X_val)
    
    # Calcular MAE
    mae = mean_absolute_error(y_val, y_pred_val)
    
    # Almacenar el mejor pipeline
    trial.set_user_attr('pipeline', pipeline)
    
    return mae


# Configurar el estudio de Optuna
study = optuna.create_study(
    direction='minimize',       # Minimizar MAE
    sampler=TPESampler(seed=42),
    study_name='xgboost_optimization'
)

# === 2. Fijar el tiempo de entrenamiento a 5 minutos (300 segundos)
start_time = time.time()

study.optimize(
    objective,
    timeout=300,
    show_progress_bar=True
)

end_time = time.time()
elapsed_time = end_time - start_time

# Corroborar tiempo
print(f"Tiempo total: {elapsed_time/60:.2f} minutos")

# === 3. Optimizar el modelo y reportar el número de *trials*, el `MAE` y los mejores hiperparámetros encontrados.

# Número de trials
print(f"Número de trials completados: {len(study.trials)}")

# Mejor MAE
print(f"\nMejor MAE en validación: {study.best_value:.4f}")

# Mejores hiperparámetros
print(f"\nMejores hiperparámetros encontrados:")
for param, value in study.best_params.items():
    print(f"  {param:20s}: {value}")

# === 5. Guardar su modelo en un archivo .pkl

# Recuperar el mejor pipeline
best_pipeline = study.best_trial.user_attrs['pipeline']

# Guardar
joblib.dump(best_pipeline, 'pipeline_xgb_optuna.pkl')

print("Modelo guardado")


   0%|          | 00:00/05:00

Tiempo total: 5.02 minutos
Número de trials completados: 163

Mejor MAE en validación: 2038.6433

Mejores hiperparámetros encontrados:
  learning_rate       : 0.07748924149980654
  n_estimators        : 849
  max_depth           : 7
  max_leaves          : 67
  min_child_weight    : 2
  reg_alpha           : 0.1706675802016543
  reg_lambda          : 0.792536761529003
  min_frequency       : 0.026944555921643847
Modelo guardado


3. **¿Cómo cambian sus resultados con respecto a la sección anterior? ¿A qué se puede deber esto?**

Los resultados mejoran respecto a la sección anterior, disminuyendo el MAE en alrededor de 400 unidades respecto al modelo XGB base y 500 respecto al modelo sesgado del punto 2. Esto puede explicarse en base a dos factores.

En primer lugar, algunos elementos del modelo óptimo mejoran el proceso de aprendizaje en comparación al modelo original. Por ejemplo, se tiene ``n_estimators = 849`` (vs default 100), donde la mayor cantidad de árboles genera un modelo más estable, ya que se obtienen más "votantes". Luego, ``max_depth = 7`` (vs default 6), genera árboles ligeramente más profundos que logran capturar patrones algo más complejos. Luego, ``reg_lambda = 0.7925`` (vs default 1) disminuye ligeramente el grado de regularización L2, permitiendo coeficientes ligeramente más grandes, lo que reduce la "savidad" del ajuste.

En segundo lugar, el modelo óptimo cuenta con diversos que ayudan a prevenir el *overfitting* y mejorar la generalización del modelo. 
Por ejemplo, ``max_leaves = 67`` (vs default 0) introduce un límite en la cantidad de hojas.
También, ``reg_alpha =  0.1707`` (vs default 0) introduce regularización L1 moderada, lo que permite al modelo, de ser necesario, realizar selección de atributos e ignorar aquellos que no aporten mucho. 
Luego, ``min_child_weight = 2`` (vs default 1), previene que se hagan divisiones que dejen pocas observaciones aisladas, lo que generaría *overfitting*.
Además, ``learning_rate = 0.0775`` (vs default 0.3), lo que disminuye la contribución de cada nuevo árbol, evitando que el modelo se "desvíe" demasiado si algún árbol se ajusta a una particularidad de los datos de entrenamiento.
Por último, ``min_frequency = 0.0270`` (vs default None) reduce ruido asociado a categorías poco frecuentes/representativas ( <2.70%) al hacer encoding, disminuyendo la dimensionalidad del modelo y mejorando la generalización.

4. **Explique cada hiperparámetro y su rol en el modelo. ¿Hacen sentido los rangos de optimización indicados?**

- 'learning_rate' (0.001, 0.1): Controla cuánto contribuye cada árbol al resultado final. El rango de optimización es conservador, fomentando un entrenamiento "lento" con más árboles, ayudando a prevenir el sobreajuste.
- 'n_estimators' (50, 1000): Número de árboles. Con más árboles, se obtiene un modelo más robusto y estable, mientras que con menos árboles se obtiene un modelo más sencillo. El límite inferior es razonable en caso de haber patrones sencillos en los datos, aunque el límite superior podría ser bajo para valores bajos de learning_rate, que tienden a necesitar más árboles.
- 'max_depth' (3, 10): Limita la profundidad máxima de cada árbol. Rango razonable, ligeramente por bajo y sobre el valor default. Valores bajo 3 serían árboles demasiado simples para la dimensionalidad del dataset, especialmente usando OneHotEncoder, mientras que sobre 10 podrían sobreajustar.
- 'max_leaves' (0, 100): Número máximo de nodos hoja por árbol. El caso mínimo corresponde al default (no se limitan hojas) lo que es razonable dado que también se restringe 'max_depth'; el máximo es difícil de interpretar, ya que no se sabe de antemano cuantos "escenarios distintos" presentan los datos. 100 podría ser razonable para "tipos" de compras de bebidas.
- 'min_child_weight' (1, 5): De la documentación, es la "Suma mínima del peso de la instancia (hessiano) necesaria en un hijo". En regresión lineal, puede entenderse como la cantidad mínima de observaciones por nodo hijo. 1 puede ser apropiado si se considera la existencia compras anómalas o particulares (por ejemplo, si se realizó una compra grupal para un evento). 5 es más conservador para prevenir *overfitting*, evitando que el modelo aprenda de estas compras "especiales".
- 'reg_alpha' (0, 1): Peso de regularización L1 (Lasso). El rango inferior genera el caso donde no se usa regularizacón L1, lo cual es razonable. Sin embargo, el rango superior es demasiado conservador, limitando la selección de atributos y, posiblemente, evitando una buena generalizacion. Podría ser aceptado dado que también se consideran otros parámetros de regularización (reg_lambda, max_depth, min_child_weight).
- 'reg_lambda' (0, 1): Peso de regularización L2 (Ridge). El rango inferior genera el caso donde no se usa regularizacón L2, lo cual es razonable. El rango superior es demasiado conservador, limitando el control de *overfitting* y, posiblemente, evitando una buena generalizacion. Podría ser aceptado dado que también se consideran otros parámetros de regularización (reg_alpha, max_depth, min_child_weight).
- 'min_frequency' (0.0, 1.0): Especifica la proporción mínima que debe representar una categoría para producir una columna en el encoding. El rango minimo representa el caso donde no se aplica la restricción, lo cual es razonable. Sin embargo, el valor máximo no es razonable pues, ni cualquier valor muy alto, pues se perdería información sobre clases que pueden ser igualmente muy representativas. Sería raro que la optimización elija un valor alto, por lo que al final es una "perdida de eficiencia".


## 4. Optimización de Hiperparámetros con Optuna y Prunners (17 puntos)

<p align="center">
  <img src="https://i.pinimg.com/originals/90/16/f9/9016f919c2259f3d0e8fe465049638a7.gif">
</p>

Después de optimizar el rendimiento de su modelo varias veces, Fiu le pregunta si no es posible optimizar el entrenamiento del modelo en sí mismo. Después de leer un par de post de personas de dudosa reputación en la *deepweb*, usted llega a la conclusión que puede cumplir este objetivo mediante la implementación de **Prunning**.

Vuelva a optimizar los mismos hiperparámetros que la sección pasada, pero esta vez utilizando **Prunning** en la optimización. En particular, usted debe:

- Responder: ¿Qué es prunning? ¿De qué forma debería impactar en el entrenamiento? [2 puntos]
- Redefinir la función `objective()` utilizando `optuna.integration.XGBoostPruningCallback` como método de **Prunning** [10 puntos]
- Fijar nuevamente el tiempo de entrenamiento a 5 minutos [1 punto]
- Reportar el número de *trials*, el `MAE` y los mejores hiperparámetros encontrados. ¿Cómo cambian sus resultados con respecto a la sección anterior? ¿A qué se puede deber esto? [3 puntos]
- Guardar su modelo en un archivo .pkl [1 punto]

Nota: Si quieren silenciar los prints obtenidos en el prunning, pueden hacerlo mediante el siguiente comando:

```
optuna.logging.set_verbosity(optuna.logging.WARNING)
```

De implementar la opción anterior, pueden especificar `show_progress_bar = True` en el método `optimize` para *más sabor*.

Hint: Si quieren especificar parámetros del método .fit() del modelo a través del pipeline, pueden hacerlo por medio de la siguiente sintaxis: `pipeline.fit(stepmodelo__parametro = valor)`

Hint2: Este <a href = https://stackoverflow.com/questions/40329576/sklearn-pass-fit-parameters-to-xgboost-in-pipeline>enlace</a> les puede ser de ayuda en su implementación

In [7]:
# Inserte su código acá

## 5. Visualizaciones (5 puntos)

<p align="center">
  <img src="https://media.tenor.com/F-LgB1xTebEAAAAd/look-at-this-graph-nickelback.gif">
</p>


Satisfecho con su trabajo, Fiu le pregunta si es posible generar visualizaciones que permitan entender el entrenamiento de su modelo.

A partir del siguiente <a href = https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/005_visualization.html#visualization>enlace</a>, genere las siguientes visualizaciones:

1. Gráfico de historial de optimización [1 punto]
2. Gráfico de coordenadas paralelas [1 punto]
3. Gráfico de importancia de hiperparámetros [1 punto]

Comente sus resultados:

4. ¿Desde qué *trial* se empiezan a observar mejoras notables en sus resultados? [0.5 puntos]
5. ¿Qué tendencias puede observar a partir del gráfico de coordenadas paralelas? [1 punto]
6. ¿Cuáles son los hiperparámetros con mayor importancia para la optimización de su modelo? [0.5 puntos]

In [8]:
# Inserte su código acá

## 6. Síntesis de resultados (3 puntos)

Finalmente:

1. Genere una tabla resumen del MAE en el conjunto de validación obtenido en los 5 modelos entrenados desde Baseline hasta XGBoost con Constraints, Optuna y Prunning. [1 punto]
2. Compare los resultados de la tabla y responda, ¿qué modelo obtiene el mejor rendimiento? [0.5 puntos]
3. Cargue el mejor modelo, prediga sobre el conjunto de **test** y reporte su MAE. [0.5 puntos]
4. ¿Existen diferencias con respecto a las métricas obtenidas en el conjunto de validación? ¿Porqué puede ocurrir esto? [1 punto]

In [9]:
# Inserte su código acá

# Conclusión
Exito!
<p align="center">
  <img src="https://i.pinimg.com/originals/55/3d/42/553d42bea9b10e0662a05aa8726fc7f4.gif">
</p>